### DISCUSSION TITLE
DSN for Caché works in pyodbc but not in SQLAlchemy

### DISCUSSION THREAD

**Discussion by Dekermanjian** [sqlalchemy/sqlalchemy#10200] [Usage Questions] [answered]
I am having a hard time getting a connection established using SQLAlchemy's `create_engine()` method using a DSN configuration. For the same DSN, I have no issues when I used pyodbc's `connect()` method. Below is how I am establishing the connection for both cases:

`create_engine("mssql+pyodbc://@mydsn")` for SQLAlchemy and
`pyodbc.connect(DSN = mydsn)` for PYODBC (this connection works)

The error message I get back from the SQLAlchemy is:
```
DBAPIError: (pyodbc.Error) ('HY000', "[HY000] [Cache ODBC][State : HY000][Native Code 359]\r\n[c:\\Users\\j.dekermanjian\\Anaconda3\\python.exe]\r\n[SQLCODE: <-359>:]\r\n[Location: ]\r\n[%msg: < User defined SQL function 'SYSTEM.SCHEMA_NAME' does not exist>] (359) (SQLExecDirectW)")
[SQL: SELECT schema_name()]
(Background on this error at: https://sqlalche.me/e/14/dbapi)
```

Any help would be really appreciated. Thank you.

**Comment by gordthompson** [comment]
From a successful connection in pyodbc, what does `connection.getinfo(6)` return?

**Reply by Dekermanjian** [reply]
Hey @gordthompson, thank you for your quick response. I don't have a method of getvalue() on my pyodbc connection object. Did you by chance mean getinfo()? If so, then I get CacheODBC3564.DLL

**Reply by gordthompson** [reply]
Yeah, I misremembered. It's `.getinfo(6)`

**Reply by Dekermanjian** [reply]
Okay, thank you for clarifying. I get CacheODBC3564.DLL

**Comment by gordthompson** [comment, marked as answer]
The Caché database is apparently not compatible with MS SQL Server. You'll need to see if a third-party SQLAlchemy dialect exists for Caché.

**Reply by Dekermanjian** [reply]
Okay, that is what I thought but was not sure because from my understanding pyodbc handles the mssql dialect in sqlalchemy so I thought it would just be a wrapper of pyodbc under the hood and just to be clear the connection works and I can run MS SQL Queries against the database in pyodbc. Thank you for your help!

**Reply by gordthompson** [reply]
`mssql+pyodbc://` does a lot of mssql-specific things, and it seems that Caché doesn't support all of them.

**Reply by Dekermanjian** [reply]
I guess I wonder if pyodbc is doing the translations of dialect under the hood to the cache database? How is this working in pyodbc?

**Reply by gordthompson** [reply]
SQLAlchemy builds the queries and pyodbc just executes them. pyodbc is a generic ODBC layer for Python.

**Reply by Dekermanjian** [reply]
Oh I see, so that is why I can run my queries using pyodbc because it uses the query exactly as I have written it. Whereas SQLAlchemy will alter my query before passing it to pyodbc. Is that correct?

**Reply by gordthompson** [reply]
In this case SQLA is sending `SELECT schema_name()` as part of initializing a connection. Once connected, you could run your own queries using SQLA's `.exec_driver_query()` or pandas' `.read_sql_query()` and SQLA wouldn't mess with them, but your connection attempt fails so you never get that far.

**Reply by Dekermanjian** [reply]
Okay, I fully understand now. Thank you very much for your help!
